### PR TITLE
Commit code for Organization.

### PR DESCRIPTION
**Pull request by wpoosanguansit**
Commit code for Organization.

### CODE REVIEW

File: organization/src/test/scala/com/improving/app/organization/utils/LoanedActorSystem.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import akka.testkit.TestKit
+import com.typesafe.config.ConfigFactory
+
+import java.util.concurrent.atomic.AtomicInteger
+import scala.jdk.CollectionConverters._
+
+trait LoanedActorSystem {
+  def configOverrides: Map[String, Any] = Map.empty
+
+  private val n = new AtomicInteger
```
Comment (reid-spencer) [inline organization/src/test/scala/com/improving/app/organization/utils/LoanedActorSystem.scala:13]: Is this even needed ? It doesn't seem referenced in this trait.

**Review by reid-spencer — COMMENTED** [review]
Please review my feedback.

File: build.sbt [hunk 2]
```diff
@@ -24,5 +26,6 @@ lazy val tenant = project.in(file("tenant"))
   .configure(C.akkaPersistentEntity("improving-app-tenant"))
   .dependsOn(commonTypes % "compile->compile;test->test")
 
-(Compile / runMain) := (tenant / Compile / runMain).evaluated
-onLoad in Global := (onLoad in Global).value andThen (Command.process("project tenant", _))
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply build.sbt]: Noted

File: organization/src/main/resources/certs/ca.pem [hunk 3]
```diff
@@ -0,0 +1,15 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/resources/certs/ca.pem:1]: I will remove this.

File: organization/src/main/resources/certs/server1.key [hunk 4]
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/resources/certs/server1.key:1]: I will remove this.

File: organization/src/main/scala/com/improving/app/organization/OrganizationServiceImpl.scala [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  }
+
+  override def getOrganization(
+      in: GetOrganizationByIdRequest
+  ): Future[Organization] = {
+    handleRequest(
+      in,
+      {
+        case StatusReply.Success(
+              response @ Organization(_, _, _, _, _, _, _, _, _)
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/OrganizationServiceImpl.scala:100]: Noted.

File: organization/src/test/scala/com/improving/app/organization/utils/LoanedActorSystem.scala (hunk 1 above)
Comment (wpoosanguansit) [reply organization/src/test/scala/com/improving/app/organization/utils/LoanedActorSystem.scala:13]: It is used in the tests using trait CassandraTestContainer.

File: organization/src/main/scala/com/improving/app/organization/domain/Organization.scala [hunk 6]
```diff
@@ -1,43 +1,815 @@
...
+          state.contacts,
+          state.meta,
+          state.status
+        )
+      )
+    }
+  }
+
+  private val commandHandler: (
+      OrganizationState,
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/domain/Organization.scala:487]: I follow the pattern Shailendra did in member and I saw that it is being a state proto message which I am using it in the handler.  We do need to have the state proto message, right?

File: organization/src/main/scala/com/improving/app/organization/utils/serialize/CborSerializable.scala [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.improving.app.organization.utils.serialize
+
+/** Marker trait for serialization with Jackson CBOR.
+  * Enabled in serialization.conf `akka.actor.serialization-bindings` (via application.conf).
+  */
+trait CborSerializable
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/utils/serialize/CborSerializable.scala:6]: Deleted.

File: organization/src/main/scala/com/improving/app/organization/utils/serialize/AvroJacksonObjectMapperFactory.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+import akka.serialization.jackson.JacksonObjectMapperFactory
+import com.fasterxml.jackson.core.JsonFactory
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.json.JsonMapper
+import com.improving.app.organization.utils.serialize.AvroJacksonObjectMapperFactory.applyObjectMapperMixins
+
+class AvroJacksonObjectMapperFactory extends JacksonObjectMapperFactory {
+
+  override def newObjectMapper(bindingName: String, jsonFactory: JsonFactory): ObjectMapper = {
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/utils/serialize/AvroJacksonObjectMapperFactory.scala:11]: Deleted.

File: organization/src/main/scala/com/improving/app/organization/utils/serialize/CborSerializable.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/utils/serialize/CborSerializable.scala:6]: Fixed.

File: organization/src/main/scala/com/improving/app/organization/domain/Organization.scala [hunk 9]
```diff
@@ -1,43 +1,815 @@
...
+  ): Behavior[OrganizationCommand] =
+    Behaviors.setup { context =>
+      context.log.info("Starting Organization {}", orgId)
+      EventSourcedBehavior
+        .withEnforcedReplies[
+          OrganizationCommand,
+          OrganizationEvent,
+          OrganizationState
+        ](
+          persistenceId = PersistenceId(entityTypeHint, orgId),
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/domain/Organization.scala:59]: Fixed.

File: organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+  def run(): Future[Http.ServerBinding] = {
+    implicit val sys = system
+    implicit val ec: ExecutionContext = system.executionContext
+
+    val service: HttpRequest => Future[HttpResponse] =
+      OrganizationServiceHandler(new OrganizationServiceImpl())
+
+    val bound: Future[Http.ServerBinding] = Http(system)
+      .newServerAt(interface = "127.0.0.1", port = 8080)
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala:99]: Fixed.

File: organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    )
+    val keyManagerFactory = KeyManagerFactory.getInstance("SunX509")
+    keyManagerFactory.init(ks, null)
+    val context = SSLContext.getInstance("TLS")
+    context.init(keyManagerFactory.getKeyManagers, null, new SecureRandom)
+    ConnectionContext.httpsServer(context)
+  }
+
+  private def readPrivateKeyPem(): String =
+    Source.fromResource("certs/server1.key").mkString
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala:82]: Deleted.

File: organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    val service: HttpRequest => Future[HttpResponse] =
+      ServiceHandler.concatOrNotFound(
+        OrganizationServiceHandler.partial(grpcService),
+        // ServerReflection enabled to support grpcurl without import-path and proto parameters
+        ServerReflection.partial(List(OrganizationService))
+      )
+
+    val bound: Future[Http.ServerBinding] = Http(system)
+      .newServerAt(interface = "127.0.0.1", port = 8080)
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala:40]: Fixed.

File: organization/src/main/resources/grpc.conf [hunk 13]
```diff
@@ -0,0 +1,11 @@
+akka.http.server.preview.enable-http2 = on
+
+organization-service {
+
+  grpc {
+    # consider setting this to a specific interface for your environment
+    interface = "0.0.0.0"
+    port = 8101
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/resources/grpc.conf:8]: Fixed.

File: organization/src/main/resources/cluster.conf [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    shutdown-after-unsuccessful-join-seed-nodes = 120s
+
+    sharding {
+      least-shard-allocation-strategy.rebalance-absolute-limit = 20
+      passivation.strategy = default-strategy
+    }
+  }
+}
+
+akka.management {
```
[the remark this answers is not in the document]
Comment (wpoosanguansit) [reply organization/src/main/resources/cluster.conf:20]: Fixed.

**Comment by wpoosanguansit** [comment]
The update on Mar 22 contains the following:

- Added organization by member and organization by owner projections
- Added specs for the projections

Fixes for feedback:

- Undo sbt project run snippet
- Remove cert and http start up
- Changed ports and Interface are variable to be set
- Changed TerminationDeadLine to be from config
- Changed match case of response to typed response
- Update extract id to be extractOrganizationId
- Changed request match cases to pass on request for processing
- Removed old serialize util

Note:

The integration test for projection does not write to the db yet.  It is being ignored for commit.
The projections are partially implemented as it is not allowed in Cassandra to delete records base on partial keys.
More work is needed to manually query the composite key data to perform the delete cleanly, ie. get all memberId from the organizations with certain orgId.

File: ddl-scripts/create-organization-tables.sql [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    billing_contacts uuid[],
+    distribution_contacts uuid[],
+    org_id uuid NOT NULL,
+    FOREIGN KEY (org_id) REFERENCES organization(org_id),
+    UNIQUE(org_id)
+);
+
+WITH insert_organization AS (
+INSERT INTO organization (org_id,
+		info_id,
```
Comment (AlexWeinstein92) [inline ddl-scripts/create-organization-tables.sql:83]: Here you are inserting `info_id` into `organization` table, but `info_id` was not declared as a field in `organization` table. The same applies to the `meta_id` field. Please ensure there are unit or integration tests to test that this is really working.

File: ddl-scripts/create-organization-tables.sql [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    name VARCHAR(255) NOT NULL,
+    short_name VARCHAR(255),
+    is_private BOOLEAN,
+    url VARCHAR(255),
+    logo VARCHAR(255),
+    tenant_id uuid,
+--    address_id uuid,
+--    FOREIGN KEY (address_id) REFERENCES address(address_id),
+--    FOREIGN KEY (tenant_id) REFERENCES tenant(tenant_id),
+    FOREIGN KEY (org_id) REFERENCES organization(org_id),
```
Comment (AlexWeinstein92) [inline ddl-scripts/create-organization-tables.sql:24]: This seems like it duplicates the `org_id` field on line 14, but I may just not be familiar with the syntax here

File: ddl-scripts/create-organization-tables.sql [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+CREATE TABLE IF NOT EXISTS public.info (
+    info_id uuid DEFAULT gen_random_uuid() PRIMARY KEY,
+    org_id uuid NOT NULL,
+    name VARCHAR(255) NOT NULL,
+    short_name VARCHAR(255),
+    is_private BOOLEAN,
+    url VARCHAR(255),
+    logo VARCHAR(255),
+    tenant_id uuid,
+--    address_id uuid,
```
Comment (AlexWeinstein92) [inline ddl-scripts/create-organization-tables.sql:21]: Why is this commented out when `address_id` is still being inserted here below?

File: ddl-scripts/create-organization-tables.sql [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    contact_id uuid DEFAULT gen_random_uuid() PRIMARY KEY,
+    primary_contacts uuid[],
+    billing_contacts uuid[],
+    distribution_contacts uuid[],
+    org_id uuid NOT NULL,
+    FOREIGN KEY (org_id) REFERENCES organization(org_id),
+    UNIQUE(org_id)
+);
+
+WITH insert_organization AS (
```
Comment (AlexWeinstein92) [inline ddl-scripts/create-organization-tables.sql:81]: I don't think we want to do this insertion in deployment.

File: organization/ddl-scripts/create_tables.cql [hunk 19]
```diff
@@ -0,0 +1,96 @@
+-- This CQL script will create the keyspace and all tables needed for the this sample.
```
Comment (AlexWeinstein92) [inline organization/ddl-scripts/create_tables.cql:1]: Please include a link to wherever this code was taken from in a comment or other documentation. Its important to keep references like that easily accessible. Please do the same for any other files adapted for our use from other projects

File: organization/src/main/scala/com/improving/app/organization/Main.scala [hunk 20]
```diff
@@ -1,14 +1,65 @@
...
+        system.terminate()
+    }
+  }
+
+  def init(implicit system: ActorSystem[_]): Unit = {
+
+    val session = CassandraSessionRegistry(system).sessionFor(
+      "akka.persistence.cassandra"
+    )
+    // use same keyspace for the item_popularity table as the offset store
```
Comment (AlexWeinstein92) [inline organization/src/main/scala/com/improving/app/organization/Main.scala:40]: This comment seems like it comes from the framework you adapted from to write this file. Please either take it out or make it up to date with your adaptations.

File: organization/src/main/scala/com/improving/app/organization/repository/OrganizationByMemberProjectionHandler.scala [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+          .map(organizations => {
+            Future.sequence(for {
+              organization <- organizations
+              member <- newMembers
+            } yield {
+              repo.updateOrganizationByMember(
+                organization.orgId.map(_.id).getOrElse("OrgId is NOT FOUND."),
+                member.id,
+                organization.copy(
+                  members = organization.members ++ newMembers,
```
Comment (AlexWeinstein92) [inline organization/src/main/scala/com/improving/app/organization/repository/OrganizationByMemberProjectionHandler.scala:55]: Most of the code here can be refactored into a single function - feel free to change the name of `handleEvent` to a better one if desired:

```
def handleEvent(newOrg: Organization) = repo
          .getOrganizationsByMemberByOrgId(orgId.id)
          .map(organizations => {
            Future.sequence(for {
              organization <- organizations
              member <- newMembers
            } yield {
              repo.updateOrganizationByMember(
                organization.orgId.map(_.id).getOrElse("OrgId is NOT FOUND."),
                member.id,
                newOrg
              )
            })
          })
          .map(_ => Done)

case MembersAddedToOrganization(Some(orgId), newMembers, actingMember, _) => {
  handleEvent(organization.copy(
    members = organization.members ++ newMembers,
    meta = Some(updateMetaInfo(organization.getMeta, actingMember))
  ))
}
```

File: ddl-scripts/create-organization-tables.sql (hunk 15 above)
Comment (wpoosanguansit) [reply ddl-scripts/create-organization-tables.sql:83]: I think I should remove this altogether as this was for the data model for relational db which is not used.  Cassandra is used for the projections.

**Comment by dag-yoppworks** [comment]
This PR will not be merged - the branch became too much to review and transfer to new developers (👋 ), and it was not close to ready for merge. There are probably useful bits to draw on in this branch, if they can be found.

This is a cautionary tale about regularly committing, reviewing, and merging your code.
alwaysbecommitting

**Pull request closed without merging**